### PR TITLE
[WIP] Retry generating Python code for `plot` and `verify`

### PR DESCRIPTION
**Pull request by xinrong-meng**
## Change
Retry generating Python code (with error messages feedbacked) for both `plot` and `verify` methods if execution fails.

We may apply the retry mechanism to other APIs in follow-ups.

## Test
Unit test and manual tests.

### CODE REVIEW

File: tests/test_ai_utils.py [hunk 1]
```diff
@@ -62,5 +62,25 @@ def test_multiple_sql_code_blocks(self):
         self.assertEqual(AIUtils.extract_code_blocks(text), ["SELECT a FROM b;", "SELECT c FROM d;"])
 
 
+class TestRetryExecution(unittest.TestCase):
```
Comment (gengliangwang) [inline tests/test_ai_utils.py:65]: @xinrong-meng can we have a test case which mocks the response  from LLM and generates bad code?

File: pyspark_ai/ai_utils.py [hunk 2]
```diff
@@ -132,3 +132,17 @@ def extract_code_blocks(text: str) -> List[str]:
...
                 return [text.strip("`")]
             return [text]
+
+    @staticmethod
+    def retry_execution(callback, max_tries=3):
+        tries = 0
+        last_exception = None
+        while max_tries > tries:
+            try:
+                return callback()
```
Comment (gengliangwang) [inline pyspark_ai/ai_utils.py:142]: We need to have a conversation with LLM and feedback the error message

Comment (xinrong-meng) [reply pyspark_ai/ai_utils.py]: Adjusted in [8ef79c0](https://github.com/pyspark-ai/pyspark-ai/pull/132/commits/8ef79c0b8c6789962a41770f3d1d08ce2d1d5ba6)

Comment (gengliangwang) [reply pyspark_ai/ai_utils.py]: @xinrong-meng could you verify the feature with gpt-3.5 or other models?

File: pyspark_ai/pyspark_ai.py [hunk 3]
```diff
@@ -484,21 +484,30 @@ def explain_df(self, df: DataFrame, cache: bool = True) -> str:
...
-        code = "\n".join(codeblocks)
-        try:
+
+        def callback(error_messages: List[str]):
+            accumulated_errors = "\n".join(error_messages)
+            instruction_base = (
+                f"The purpose of the plot: {desc}." if desc is not None else ""
+            )
+            instruction = (
+                f"{instruction_base}\nPlease avoid errors:\n{accumulated_errors}"
```
Comment (gengliangwang) [inline pyspark_ai/pyspark_ai.py:495]: TBH I prefer a clean way as follows:
```
SystemMessage("you are an assistant on plotting...")
HumanMessage("plot 4 week moving average")
AIMessage("import pandas as pd...")
HumanMessage("Error:...")
AIMessage("....")
```

**Comment by gengliangwang** [comment]
I am closing this one since we have https://github.com/pyspark-ai/pyspark-ai/pull/159

**Pull request closed without merging**